### ISSUE TITLE
Change Request: Support regular expressions as option values

### ISSUE THREAD

**Issue by DMartens** [eslint/eslint#18648]
### ESLint version

v9.6.0

### What problem do you want to solve?

Currently there is no way to specify rule options schema validating that a value is a regular expression.
But there is a need for it as various rules (e.g. core rules [id-match](https://github.com/eslint/eslint/blob/6e79ac76f44b34c24a3e92c20713fbafe1dcbae2/lib/rules/id-match.js#L62) or no-fallthrough) allow users to pass a string which is used to instantiate a regular expression.

Advantages for users:
- known that the value is handled as a regular expression
- allows passing regular expression flags
- proper syntax highlighting
- lintable 

Advantages for rule developers:
- no need to verify that the string is a valid regular expression
- no custom parsing (e.g. some rules use surrounding `/` to check whether the string should be handled as a regular expression or for regular expression flags)

Also passing regular expressions is possible in the flat config by providing an empty schema `{}` and the rule manually checking whether the value is a regular expression:
```javascript
export default {
	meta: {
		schema: {
			type: 'object',
			properties: {
				regex: {}
			}
		}
	},
	create(context) {
		if (!context.options[0].regex instanceof RegExp) {
			return {};
		}
		...
	}
}
```

### What do you think is the correct solution?

ajv can be [extended with new keywords](https://ajv.js.org/keywords.html) which can providate a validator for regular expressions.
Unfortunately ajv does not provide a way to extend the `type` keyword.
A simple implementation would be:
```javascript
// Adds { regexp: true } to verify that the passed value is a regular expression
ajv.addKeyword({
  keyword: "regexp",
  schema: false,
  validate(data) => data instanceof RegExp
})
```

### Participation

- [X] I am willing to submit a pull request for this change.

**Comment by DMartens** [comment]
If that is the only concern, then we could disallow the global flag for regular expressions.
As `lastIndex` is the only property about a regular expression which is not serializable.
